Exponential Histogram implementation

sritchie opened the pull request:
This PR implements the exponential histogram, or DGIM, algorithm from this paper: http://ilpubs.stanford.edu:8090/504/1/2001-34.pdf

The code for `ExpHist` and `ExpHist.Canonical` have a good description of what's going on and how the algorithm works. The supplied `tut`-based documentation has a description of the l-canonical representation procedure.

### Next Steps
- [ ] Add a Monoid implementation based on the algorithm presented here http://megaslides.com/doc/1974/ecm-sketches
- [ ] wrap this impl with a queue that buffers N items before pushing them all into the `ExpHist`, and update relevant error calculations.
- [ ] Get some benchmarks going. I'm going to propose that we do this in the next PR.

cc @non @johnynek

sritchie commented on algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala:
```diff
@@ -0,0 +1,342 @@
+   * of 2 matches `s`'s l-canonical representation (for the supplied
+   * l).
+   */
+  def bucketsFromLong(s: Long, l: Int): Vector[Long] = {
```
@non @johnynek would love some advice on how to efficiently share the calculations going on here and in `fromLong`, without creating an intermediate return tuple for the values that both functions need.

That said... `bucketsFromLong` is the only thing we're actually using in the implementation. I could just rebuild `fromLong` from the results of `bucketsFromLong`.

sritchie replied:
this is exactly `toBuckets(fromLong(s, k))` btw

oscar-stripe replied:
can you add that comment: `bucketsFromLong(s, l) == toBuckets(fromLong(s, k))`

This is a different representation right? Can we have a different AnyVal wrapper to distinguish?

sritchie commented on algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala:
```diff
@@ -0,0 +1,342 @@
+      total = total + delta)
+  }
+
+  def oldestBucketSize: Long = if (total == 0) 0L else buckets.last.size
```
I wonder if it'd be a good idea to skip the zero checking by adding a `trait ExpHist` and a `case object EmptyExpHist`.

sritchie commented on algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala:
```diff
@@ -0,0 +1,342 @@
+   * window of size `windowSize`.
+   */
+  case class Config(epsilon: Double, windowSize: Long) {
+    val k: Int = math.ceil(1 / epsilon).toInt
```
we never really need `k`, but it shows up in the paper.

sritchie commented on algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala:
```diff
@@ -0,0 +1,342 @@
+    val (b @ Bucket(count, _)) +: tail = input
+    (toDrop - count) match {
+      case 0 => tail
+      case x if x < 0 => b.copy(size = -x) +: tail
```
this is the only reason I'm not able to use an iterator, this pushing back into the vector. I wonder if we can implement this better with some carry Option. thoughts?

sritchie commented on algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala:
```diff
@@ -0,0 +1,342 @@
+      if (delta == 0)
+        step(timestamp)
+      else {
+        addAllWithoutStep(sorted, delta).step(timestamp)
```
this was a little subtle. We have to step AFTER adding because the range of timestamps of the new items we're pushing in might be larger than windowSize.

sritchie commented on algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala:
```diff
@@ -0,0 +1,342 @@
+  def add(delta: Long, timestamp: Long): ExpHist = {
+    val self = step(timestamp)
+    if (delta == 0) self
+    else self.addAllWithoutStep(Vector(Bucket(delta, timestamp)), delta)
```
here we can step BEFORE, because we know that there's only a single timestamp coming in.

oscar-stripe commented on algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala:
```diff
@@ -0,0 +1,342 @@
+  def guess: Double =
+    if (total == 0) 0.0
+    else (total - (oldestBucketSize - 1) / 2.0)
+
```
can you add a `Approximate[Long]` return value? Those compose somewhat.

sritchie replied:
boom, will do

sritchie replied:
don't we want `approximate[Double]`, since we want the estimate to fall in the middle of the range?

sritchie replied:
okay, added!

oscar-stripe commented on algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala:
```diff
@@ -0,0 +1,342 @@
+  @inline private[this] def floorPowerOfTwo(x: Long): Int =
+    JLong.numberOfTrailingZeros(JLong.highestOneBit(x))
+
+  @inline private[this] def modPow2(i: Int, exp2: Int): Int = i & ((1 << exp2) - 1)
```
what about the bugs on inlining you found? Are we sure none of these exhibit this? Downstream folks still could if they compile with `-optimize`.

oscar-stripe commented on algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala:
```diff
@@ -0,0 +1,342 @@
+   * - ret(i) for all i < j == l or l + 1
+   * - ret(j) < l + 1
+   */
+  def fromLong(s: Long, l: Int): Vector[Int] = {
```
no curlies...

oscar-stripe replied:
what about `final case class CanonicalRepresentation(toVector: Vector[Int]) extends AnyVal`

Also, maybe add these specialized types under `object ExpHist`.

sritchie replied:
removed

sritchie replied:
I like it. I moved the whole object inside `ExpHist` - will add these as well.

oscar-stripe commented on algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala:
```diff
@@ -0,0 +1,342 @@
+   *
+   * the "l" means that
+   *
+   * - ret(i) for all i < j == l or l + 1
```
what is `ret`? The returned array? Can we comment to that?

sritchie replied:
fixed up

oscar-stripe commented on algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala:
```diff
@@ -0,0 +1,342 @@
+   * (i = vector index, j = index of last entry)
+   *
+   * returns a vector of the the coefficients of s^i in the
+   * l-canonical representation of s.
```
this confuses me. If you are returning the coefficients of `s^i` in the `i'th` position of the vector, then why isn't the vector infinite? We can take infinite powers?

sritchie replied:
huge typo! should have been 2^i. Fixed up the docs too.

oscar-stripe commented on algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala:
```diff
@@ -0,0 +1,342 @@
+   * @param rep l-canonical representation of some number s for some l
+   * @return The original s
+   */
+  def toLong(rep: Vector[Int]): Long =
```
can this be a method on an `AnyVal` class?

sritchie replied:
boom, moved that and `toBuckets`

oscar-stripe commented on algebird-test/src/test/scala/com/twitter/algebird/ExpHistLaws.scala:
```diff
@@ -0,0 +1,281 @@
+    }
+  }
+
+  def isPowerOfTwo(i: Long): Boolean = (i & -i) == i
```
fancy! :) Hacker's Delight in the house.

sritchie commented:
Btw the test failure in the `update paper link` push is fixed by the sbt-mima-plugin bump to 0.1.11 in https://github.com/twitter/algebird/pull/556.

johnynek commented on algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala:
```diff
@@ -0,0 +1,405 @@
+      else {
+        rep.iterator.zipWithIndex
+          .map { case (i, exp) => i.toLong << exp }
+          .reduce(_ + _)
```
there is going to be a fair amount of boxing cost here. Actually calling `Monoid.sum( )` will avoid some of that since `Monoid` is specialized on `Long`

sritchie replied:
how about

```scala
Monoid.sum {
        rep.iterator.zipWithIndex
          .map { case (i, exp) => i.toLong << exp }
      }
```

johnynek replied:
👍

johnynek commented on algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala:
```diff
@@ -0,0 +1,405 @@
+     * @return vector of powers of 2 (where ret.sum == the original s)
+     */
+    def toBuckets: Vector[Long] =
+      rep.zipWithIndex.flatMap { case (i, exp) => List.fill(i)(1L << exp) }
```
what about:
```scala
rep.iterator
  .zipWithIndex
  .flatMap { case (i, exp) => Iterator.fill(i)(1L << exp) }
  .toVector
```

sritchie replied:
done

johnynek commented on algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala:
```diff
@@ -0,0 +1,405 @@
+   * into this exponential histogram instance.
+   */
+  def fold: Fold[Bucket, ExpHist] =
+    Fold.foldLeft(this) {
```
I wonder about using a `foldMutable` here and accumulating a batch of `Bucket` and then doing an `addAll` which might be a lot faster, but we can punt on this if you like.

sritchie replied:
added a test and this impl:

```scala
def fold: Fold[Bucket, ExpHist] =
    Fold.foldMutable[Builder[Bucket, Vector[Bucket]], Bucket, ExpHist](
      { case (b, bucket) => b += bucket },
      { _ => Vector.newBuilder[Bucket] },
      { x => addAll(x.result) })
```

way better imo.

sritchie commented on algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala:
```diff
@@ -0,0 +1,405 @@
+   *  the returned ExpHist will have the same timestamp, equal to
+   *  `ts`.
+   */
+  def from(i: Long, ts: Long, conf: Config): ExpHist = {
```
yeah, let me thread it through and see what it looks like

sritchie commented on algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala:
```diff
@@ -0,0 +1,405 @@
+  def addAll(unsorted: Vector[Bucket]): ExpHist =
+    if (unsorted.isEmpty) this
+    else {
+      val sorted = unsorted.sorted(Ordering[Bucket].reverse)
```
done

sritchie commented:
Okay @johnynek, addressed all comments. I think we should add the Monoid in the next round, so we can experiment with controlling the relative error. We'll probably need to track the error separately from the calculation I have here...

We might need to create an `ApproximateHistogram` trait that we can use for the `ExpHist`, the `Windows` implementation, the `ZeroExpHist` and the `ExpHist` that keeps a queue of items around before adding.

sritchie commented:
Amazing, the docs build failed after my change!! Fixing now.

oscar-stripe commented:
So good! Love tut docs! Thanks for an excellent example for us to follow @sritchie !

👍 

Any concerns @isnotinvain ?

sritchie commented:
Thanks for the review, @johnynek! @isnotinvain, would love a final sign off.

johnynek commented:
I think the rules say it is fine to merge now.
